File: PerfectSquares4.cpp
#include<iostream>
#include<math.h>
#include<algorithm> 
#include<vector>

using namespace std;

class Solution {
public:
    int numSquares(int n) {
        
        vector<int> dp(1, 0);
        
        for(int i = 1;i<=n;i++) {
        	int val = INT_MAX;
        	for(int j=1;j*j<=i;j++) {
        		val = min(val, dp[i - j * j] + 1);
			}
			dp.push_back(val);
		}
		
		return dp.back();
        
    }
    
};

int main() {
	
	int n;
	cin>>n;
	
	Solution *solution = new Solution();
	cout<<solution->numSquares(n);
	
	return 0;
	
}

File: PerfectSquares4.py
import math
class Solution:
    def numSquares(self, n):
        """
        :type n: int
        :rtype: int
        """
        while n % 4 == 0:
            n //= 4
        if n % 8 == 7:
            return 4
        a = 0
        while a * a <= n:
            b = int(math.sqrt(n - a * a))
            if b * b == n - a * a:
                if b == 0 or a == 0:
                    return 1
                else:
                    return 2
            a += 1
        return 3

if __name__ == '__main__':
    n = int(input())
    solution = Solution()
    print(solution.numSquares(n))
